src/tracing/esp32_diagnostic_trace/DiagnosticTracing.cpp

```
@@ -0,0 +1,169 @@
+
+void ESP32Diagnostics::LogNodeDiscoveryFailed(NodeDiscoveryFailedInfo & info) {}
+
+void ESP32Diagnostics::LogMetricEvent(const MetricEvent & event)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    switch (event.ValueType())
+    {
+    case ValueType::kInt32: {
+        ESP_LOGI("mtr", "The value of %s is %ld ", event.key(), event.ValueInt32());
+        Diagnostic<int32_t> metric(event.key(), event.ValueInt32(), esp_log_timestamp());
+        err = mStorageInstance.Store(metric);
+    }
+    break;
+
+    case ValueType::kUInt32: {
+        ESP_LOGI("mtr", "The value of %s is %lu ", event.key(), event.ValueUInt32());
+        Diagnostic<uint32_t> metric(event.key(), event.ValueUInt32(), esp_log_timestamp());
+        err = mStorageInstance.Store(metric);
+    }
+    break;
+
+    case ValueType::kChipErrorCode:
+        ESP_LOGI("mtr", "The value of %s is error with code %lu ", event.key(), event.ValueErrorCode());
+        break;
+
+    case ValueType::kUndefined:
+        ESP_LOGI("mtr", "The value of %s is undefined", event.key());
+        break;
+
+    default:
+        ESP_LOGI("mtr", "The value of %s is of an UNKNOWN TYPE", event.key());
+        break;
+    }
+
+    VerifyOrReturn(err == CHIP_NO_ERROR, ChipLogError(DeviceLayer, "Failed to store Metric Diagnostic data"));
+}
+
+void ESP32Diagnostics::TraceCounter(const char * label)
+{
+    ESPDiagnosticCounter::GetInstance(label).ReportMetrics(label, mStorageInstance);
+}
+
+void ESP32Diagnostics::TraceBegin(const char * label, const char * group)
+{
+    if (IsPermitted(group))
+    {
+        StoreDiagnostics(label, group);
+    }
+}
+
+void ESP32Diagnostics::TraceEnd(const char * label, const char * group) {}
+
+void ESP32Diagnostics::TraceInstant(const char * label, const char * value)
+{
+    if (!IsPermitted(value))
+    {
+        StoreDiagnostics(label, value);
+    }
+}
+
+void ESP32Diagnostics::StoreDiagnostics(const char * label, const char * group)
```

should we propagate the error a level up?